### PR TITLE
feat: add counter detection logic for nic health monitor

### LINKED ISSUE #1212: [Feature]: Add counter detection logic in NIC health monitor
### Prerequisites

- [x] I searched existing issues

### Feature Summary

Add counter detection logic in NIC health monitor as outlined in: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-counter-detection.md

### Problem/Use Case

Add counter detection logic in NIC health monitor as outlined in: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-counter-detection.md

### Proposed Solution

Refer to the doc: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-counter-detection.md

### Component

Health Monitor

### PR DESCRIPTION
**Pull request by KaivalyaMDabhadkar**
## Summary

Design doc for changes: [Preview](https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-counter-detection.md)

## Tests:

The changes have been tested with tilt tests as well which are all passing. The tilt tests will be merged in a subsequent PR.

## Type of Change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [ ] Fault Management
- [x] Health Monitors
- [ ] Janitor
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [x] Documentation updated (if needed)
- [x] Ready for review

## Summary by CodeRabbit

* **New Features**
  * Counter-based degradation checks for InfiniBand and Ethernet NICs with latching, reset/recovery, and delta/velocity evaluation; counter checks run on a dedicated 1s polling loop.

* **Configuration**
  * New counterDetection block with per-counter thresholds, velocity units, fatality flags, and recommended actions; degradation checks enabled by default; configuration now supplied as a TOML ConfigMap.

* **Metrics**
  * Added Prometheus metric for counter threshold breaches.

* **Documentation**
  * Docs updated for 1s polling, counter semantics, latching/recovery, and config examples.

* **Tests**
  * Comprehensive unit tests covering counter evaluation logic and edge cases.

### CODE REVIEW

File: docs/nic-health-monitor/link-counter-detection.md [hunk 1]
```diff
@@ -1164,89 +1140,93 @@ counterDetection:
 
 ### 10.4 Threshold Processing Algorithm
 
+The evaluator uses **latching breach** semantics: once a counter
+breaches its threshold, the breach flag stays set until the counter is
+reset (`current < previous`) or the host reboots. Polls while a counter
+is already breached emit nothing; recovery events fire only on counter
+reset of a previously breached counter.
+
 ```
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply docs/nic-health-monitor/link-counter-detection.md:1132]: Not necessary

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+		CheckName: effectiveCheckName,
+		IsFatal:   counterCfg.IsFatal,
+		Since:     now,
+	}
+
+	message := fmt.Sprintf(
+		"Port %s port %s: %s - %s (value=%d, delta=%d, rate=%.2f/sec)",
+		device, port, counterCfg.Name, counterCfg.Description,
+		currentValue, delta, rate,
+	)
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go:432]: Fixed

File: health-monitors/nic-health-monitor/pkg/sysfs/reader.go [hunk 3]
```diff
@@ -92,6 +92,23 @@ func (r *fsReader) ReadNetOperState(iface string) (string, error) {
...
+}
+
+func readUint64(path string) (uint64, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read %s: %w", path, err)
+	}
+
+	return strconv.ParseUint(strings.TrimSpace(string(data)), 10, 64)
+}
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/sysfs/reader.go:115]: Fixed

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+
+// ComputeRate is the exported form of computeRate for tests.
+func ComputeRate(delta uint64, elapsed time.Duration, unit string) float64 {
+	return computeRate(delta, elapsed, unit)
+}
+
+// ResolveAction maps a config string to the protobuf enum. Only the
+// two values that appear in the default counter configs are recognised;
+// everything else falls through to NONE.
+func ResolveAction(action string) pb.RecommendedAction {
```
Comment (XRFXLP) [inline health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: why this value is exposed through config if it is not freely configurable? If we only want two recommended action would exposing just `isFatal: true` with underlying assumption that it translates to reboot, works?

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+
+		if event := e.evaluateOne(key, currentValue, now, counterCfg, entities, checkName); event != nil {
+			events = append(events, event)
+		}
+	}
+
+	return events
+}
+
+// EvaluateNetCounters reads interface-level counters (currently only
```
Comment (XRFXLP) [inline health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: should this be validated during startup that we only support one fixed path even though free configuration is allowed? Otherwise we're silently skipping anything other than `statistics/carrier_changes`

File: health-monitors/nic-health-monitor/pkg/checks/counter/ethdegradation.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+// Name returns the check identifier.
+func (c *EthernetDegradationCheck) Name() string {
+	return checks.EthernetDegradationCheckName
+}
+
+// Run executes a single Ethernet/RoCE counter degradation poll.
+func (c *EthernetDegradationCheck) Run() ([]*pb.HealthEvent, error) {
+	if !c.cfg.CounterDetection.Enabled {
```
Comment (lalitadithya) [inline health-monitors/nic-health-monitor/pkg/checks/counter/ethdegradation.go]: Can this check be based upstream?

File: health-monitors/nic-health-monitor/pkg/checks/counter/ibdegradation.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+// Name returns the check identifier.
+func (c *InfiniBandDegradationCheck) Name() string {
+	return checks.InfiniBandDegradationCheckName
+}
+
+// Run executes a single InfiniBand counter degradation poll. After
+// evaluation it merges the evaluator's snapshots and breach flags back
+// into the shared state file and persists if either changed.
+func (c *InfiniBandDegradationCheck) Run() ([]*pb.HealthEvent, error) {
+	if !c.cfg.CounterDetection.Enabled {
```
Comment (lalitadithya) [inline health-monitors/nic-health-monitor/pkg/checks/counter/ibdegradation.go]: Can this check be made upstream to this?

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+// ComputeRate is the exported form of computeRate for tests.
+func ComputeRate(delta uint64, elapsed time.Duration, unit string) float64 {
+	return computeRate(delta, elapsed, unit)
+}
+
+// ResolveAction maps a config string to the protobuf enum. Only the
+// two values that appear in the default counter configs are recognised;
+// everything else falls through to NONE.
+func ResolveAction(action string) pb.RecommendedAction {
+	if action == "REPLACE_VM" {
```
Comment (lalitadithya) [inline health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: I believe there should be a protobuf enum to map string to do this mapping

File: health-monitors/nic-health-monitor/pkg/checks/counter/ethdegradation.go (hunk 6 above)
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/ethdegradation.go]: Done, moved it to buildChecks in main.go

File: health-monitors/nic-health-monitor/pkg/checks/counter/ibdegradation.go (hunk 7 above)
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/ibdegradation.go]: Done, moved it to buildChecks in main.go

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go (hunk 4 above)
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: Good point, removed recommendedAction from the config entirely. ResolveAction now derives the action from isFatal directly (true -> REPLACE_VM, false -> NONE).

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go (hunk 5 above)
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: Good catch, Fixed by changing the routing to use a statistics/ prefix match instead. Also added config validation that rejects paths not starting with counters/, hw_counters/, or statistics/

File: health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go (hunk 8 above)
Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: I have now removed recommendedAction as a config field entirely according to this: https://github.com/NVIDIA/NVSentinel/pull/1248#discussion_r3186501668

Comment (KaivalyaMDabhadkar) [reply health-monitors/nic-health-monitor/pkg/checks/counter/evaluator.go]: No string-to-enum mapping needed anymore

**Review by lalitadithya — APPROVED** [review]
